### PR TITLE
W-17440028 - Fix FIPS validation

### PR DESCRIPTION
**Pull request by ppravallika9**
Changes include:
1. Update in xmlsec version for FIPS compliance
2. Unused bouncy castle dependencies have been removed
3. Tests have been modified to be able to run in FIPS mode, using valid bcfks keys

_Hi, thank you for your work. We understand that you want to merge your changes and move on from this issue, but we also want to maintain the safety, readability, and testability of our code. Because of this, we kindly ask that you confirm that you have fulfilled the prerequisites for each category of activity before merging your changes._

### If the Pull Request has a dependency update:

- [ ] I have read the Release Notes for the new version (and intermediate versions if the jump would include more than one). _Don't blindly trust that the dependency will honor SEMVER, always read carefully the release notes_
- [ ] Java 8+ support is maintained
- [ ] The new version has no vulnerabilities
- [ ] A team member reviewed the Pull Request.
- [ ] No backward compatibility is broken.
- [ ] If it is a "provided" dependency, it has been checked that if it is suggested, the version number is the same.

### For ALL the Releases:

- [ ] There is a task in GUS for this change.
- [ ] The corresponding Release Notes have been written and shared with the documentation team.
- [ ] The new code has tests and they have before and after methods to be sure that you are working with a clean environment and that you are leaving the environment clean after they finish.
- [ ] If you are using reflection, create a test to call the methods you are invoking, this way, if there is a change in the API, the test will be able to detect it.
- [ ] If the project has a parent pom and modules, the release pipeline only uploads the modules, so, please, make sure that the parent was deployed in the repositories before considering this release finished.
- [ ] Notify in the Slack Channel that the release is complete, so the Docs team can merge the Documentation and Release Notes PR. The docs team always checks in Anypoint Exchange that the connector is released before merging the release notes.
- [ ] Add the connector build to GUS, and assign that build to the GUS ticket.
- [ ] Create a change case request, based on the Change Case Management doc https://docs.google.com/document/d/1tMJlTTZLaXMLiYwxlMBFY1yJwBmejhc4-IP8HAXgDfY/edit#

**NOTE**: _(applies only for Core-Connectors connectors and modules) The release process ends when you merge the pull request that updates the support branch to the new snapshot, please don't forget to do this._

### Patch Version:

- [ ] The Pull Request has been peer-reviewed.
- [ ] No backward compatibility is broken.
- [ ] Documentation: (Required) Release Notes PR with compatibility table. Share it with the Docs team.

### Minor Version:

- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-1.1.1-new-chiper) to coordinate the release of this version with the documentation team and the architects of our team (at least one week before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been done and has been updated with the changes related to the new functionalities.
- [ ] The pull request has been reviewed by a peer and the team leader.
- [ ] No backward compatibility is broken.
- [ ] Public Documentation: Minor releases nearly always require a revision of the documentation because the release usually includes new features as well as bug fixes. Even if there are no user-facing changes, documentation has to be versioned for every minor release.
  - [ ] (Required) Release Notes PR with compatibility table. Share it with the Docs team.
  - [ ] (Likely required) Reference guide. Run the script and share the file with the Docs team.
  - [ ] (Assess) Often a minor release impacts UI and examples, so the documentation (user guides, examples, troubleshooting guides) should be reviewed and updated accordingly. Notify the docs team if any updates are needed.

### Major Version:

- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-2.0.0) to coordinate the release of this version with the documentation team and the architects of our team (at least one month before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been made and updated with the changes related to the new functionalities and possible compatibility problems with the previous version.
- [ ] The systems properties have been removed, detailing in the documentation the changes in the default behavior of the product.
- [ ] The TODO comments of the code have been reviewed.
- [ ] A colleague, the team leader, and the team architect have reviewed the Pull Request.
- [ ] Public Documentation: Major releases break backward compatibility and should never be released without documentation, as this breaks the user experience.
  - [ ] (Required) Upgrade and Migration Guide: Every major release must have a detailed upgrade guide (template) with information about what was changed and what steps the user has to take to ensure the connector works as expected. Share the details with the Docs team so they can update the guide.
  - [ ] (Required) Release Notes PR with compatibility table. Share it with the docs team.
  - [ ] (Required) Reference guide. Run the script and share the file with the Docs team.
  - [ ] User guide updates, new screenshots or examples (if applicable), share the updates with the docs team.

### CODE REVIEW

**Review by sabecasissf — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 1]
```diff
@@ -551,5 +587,63 @@
...
+            <id>fips-config</id>
+            <activation>
+                <property>
+                    <name>mule.security.model</name>
+                    <value>fips140-2</value> <!-- if mule.security.model == fips140-2 -->
+                </property>
+            </activation>
+            <properties>
+                <env.docker.soapui.container.project.folder>soapui-prj-fips</env.docker.soapui.container.project.folder>
+                <org.apache.xml.security.securerandom.algorithm>DRBG</org.apache.xml.security.securerandom.algorithm>
```
Comment (sabecasissf) [inline pom.xml:628]: I believe setting the org.apache.xml.security.securerandom.algorithm property on the current xmlsec version is no longer necessary. Please try removing it from the pom file and check if all tests are passing.

File: src/test/munit/wss/encrypt-test-case.xml [hunk 2]
```diff
@@ -45,11 +45,26 @@
         </wsc:connection>
     </wsc:config>
 
+    <wsc:config name="encrypt-config-fips">
+        <wsc:connection wsdlLocation="http://localhost:${wsPort}/${soapVersion}/?wsdl" service="TestService" port="TestPort" address="http://localhost:${wsPort}/${soapVersion}/">
+            <wsc:web-service-security>
+                <wsc:encrypt-security-strategy>
+                    <wsc:key-store-configuration keyStorePath="security/bcfks/ssltest-cacerts.bcfks" password="changeit" alias="s1as" type="bcfks"/>
+                    <wsc:encryption-algorithms-configuration encryptionSymAlgorithm="AES_256_GCM" encryptionDigestAlgorithm="SHA256">
```
Comment (sabecasissf) [inline src/test/munit/wss/encrypt-test-case.xml:53]: Does this work on the FIPS pipeline if you leave it as encryptionSymAlgorithm=AES_128 and encryptionDigestAlgorithm="SHA1" ?

**Comment by sabecasissf** [comment]
@ppravallika9  Nice job fixing the keys for the encryption tests. Can you please add a brief explanation of what was changed in the PR description? In addition, since there are library version changes directly related to the FIPS work, those are compilation changes and must be documented in a Technical Specification. Can you please create it and add the link to it in the description of this PR as well?
Thank you!

**Comment by sabecasissf** [comment]
Sorry, I meant to add a comment and closed it by mistake. Reopened now.

File: src/test/munit/mtom/mtom-and-signature-verification-test-case.xml [hunk 3]
```diff
@@ -31,6 +31,21 @@
         </wsc:connection>
     </wsc:config>
 
+    <wsc:config name="sign-mtom-config-fips">
+        <wsc:connection wsdlLocation="http://localhost:${WS_SOAP11_PORT}/SOAP11/?wsdl"
```
Comment (sabecasissf) [inline src/test/munit/mtom/mtom-and-signature-verification-test-case.xml:35]: @ppravallika9 do you consider any of this test changes applicable for the creation of a migration guide or clarification on the connector documentation under FIPS mode? If so, please create a task for the documentation team with the proper changes to document.

Comment (ppravallika9) [reply src/test/munit/mtom/mtom-and-signature-verification-test-case.xml:35]: Since we would be releasing a new minor for this connector for FIPS, updating the docs would be tracked for release in the below task
https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000029jsGcYAI/view

File: src/test/munit/wss/encrypt-test-case.xml (hunk 2 above)
Comment (ppravallika9) [reply src/test/munit/wss/encrypt-test-case.xml:53]: This works with both "AES_128" and "AES_256_GCM" in FIPS mode

File: pom.xml [hunk 4]
```diff
@@ -56,19 +55,42 @@
         <mavenResourcesVersion>3.3.1</mavenResourcesVersion>
 
         <!-- Plugin dependencies -->
-        <apache.santuario.xmlsec.version>2.3.4</apache.santuario.xmlsec.version>
+        <apache.santuario.xmlsec.version>3.0.4</apache.santuario.xmlsec.version>
```
Comment (acurci-at-mulesoft) [inline pom.xml:58]: Hi, @ppravallika9 . I don't remember why this connector was failing. Could you remember me that? Why do we need upgrade apache santuario? Did you check the release notes?

File: pom.xml [hunk 5]
```diff
@@ -56,19 +55,42 @@
...
+        <config.ref.name.invalid-encrypt-config>${env.config.name.invalid-encrypt-config}</config.ref.name.invalid-encrypt-config>
+        <config.ref.name.decrypt-config>${env.config.name.decrypt-config}</config.ref.name.decrypt-config>
+        <config.ref.name.timestamp-config>${env.config.name.timestamp-config}</config.ref.name.timestamp-config>
+        <config.ref.name.security-header-order-default-config>${env.config.name.security-header-order-default-config}</config.ref.name.security-header-order-default-config>
+        <config.ref.name.security-header-timestamp-first-config>${env.config.name.security-header-timestamp-first-config}</config.ref.name.security-header-timestamp-first-config>
+        <config.ref.name.security-header-timestamp-last-config>${env.config.name.security-header-timestamp-last-config}</config.ref.name.security-header-timestamp-last-config>
+        <config.ref.name.decryptAndTimestampOnResponse-config>${env.config.name.decryptAndTimestampOnResponse-config}</config.ref.name.decryptAndTimestampOnResponse-config>
+        <config.ref.name.decryptOnResponse-config>${env.config.name.decryptOnResponse-config}</config.ref.name.decryptOnResponse-config>
```
Comment (acurci-at-mulesoft) [inline pom.xml:85]: Just as an alternative approach, another way to achieve the same result could be by using a postfix and concatenating it whenever the value is needed.

This way, you would only need to define two values:

If FIPS → postfix = "-fips"
If Non-FIPS → postfix = ""
Then, in the settings, you could use config="example${postfix}", which would dynamically resolve to either example or example-fips, depending on the case.

Just a thought—let me know what you think!

File: src/test/java/org/mule/extension/ws/ConfigEqualsTestCase.java [hunk 6]
```diff
@@ -76,4 +79,44 @@ public void testCustomTransportsEquals() {
...
+    anotherInstance.setEncryptSecurityStrategy(mockEncryptSecurityStrategy);
+    anotherInstance.setIncomingTimestampSecurityStrategy(mockIncomingTimestampSecurityStrategy);
+    anotherInstance.setActor("http://example.com");
+    anotherInstance.setMustUnderstand(true);
+
+    assertTrue(webServiceSecurity.equals(anotherInstance));
+    assertEquals(webServiceSecurity.hashCode(), anotherInstance.hashCode());
+
+    anotherInstance.setActor("http://changed.com");
+    assertNotEquals(webServiceSecurity, anotherInstance);
```
Comment (acurci-at-mulesoft) [inline src/test/java/org/mule/extension/ws/ConfigEqualsTestCase.java:120]: shouldn't you check `assertNotEquals(webServiceSecurity.hashCode(), anotherInstance.hashCode());`?

File: src/test/java/org/mule/extension/ws/internal/transport/CursorStreamWithProviderTest.java
Comment (acurci-at-mulesoft) [inline src/test/java/org/mule/extension/ws/internal/transport/CursorStreamWithProviderTest.java:1]: It would be helpful to add tests covering exception handling and null values.

File: src/test/resources/security/bcfks/clientKeyStore.bcfks
Comment (acurci-at-mulesoft) [inline src/test/resources/security/bcfks/clientKeyStore.bcfks:1]: I see that all the keystores are stored in the resources folder... I think the ideal thing would be to take them to Vault instead of having them stored in the repository, don't you think? Could you create a user story for that connector to do that refactoring in the future?

**Review by acurci-at-mulesoft — COMMENTED** [review]
Please check my comments

File: pom.xml (hunk 4 above)
Comment (ppravallika9) [reply pom.xml:58]: The older version was giving SecureRandom wrong algorithm error. Below is the issue description and fix has been made from 3.0.1: https://issues.apache.org/jira/browse/SANTUARIO-589

File: src/test/java/org/mule/extension/ws/ConfigEqualsTestCase.java (hunk 6 above)
Comment (ppravallika9) [reply src/test/java/org/mule/extension/ws/ConfigEqualsTestCase.java]: This is now fixed

File: src/test/java/org/mule/extension/ws/internal/transport/CursorStreamWithProviderTest.java
Comment (ppravallika9) [reply src/test/java/org/mule/extension/ws/internal/transport/CursorStreamWithProviderTest.java:1]: I have now updated with few tests covering null values

File: pom.xml (hunk 5 above)
Comment (ppravallika9) [reply pom.xml:85]: since we are defining common properties first, and then the profiles. Even if we set "postfix" based on profile it would still needed to added twice to the profile based properties (something like this sign-mtom-config${postfix} would appear in both the profiles). I can update this, but the number of lines would still remain the same

File: src/test/resources/security/bcfks/clientKeyStore.bcfks
Comment (ppravallika9) [reply src/test/resources/security/bcfks/clientKeyStore.bcfks:1]: Logged a work item to track this: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000029yRVzYAM/view

**Review by acurci-at-mulesoft — APPROVED** [review]
